history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
> Assume at epoch t, Alice send a registration request, then Alice's key would be inserted to the tree at epoch t+1. It means the Alice's key should not be found in the tree at epoch t but at epoch t+1, right?

Correct.

> But if we do as above, Alice's key would be found in the tree at epoch t (since we inserted to the tree before cloning it). 
> 
> Do I have any mistakes? 

I think you're right. For @arlolra's suggestion to work in the same way we think tree updates should work, the inserts into the tree that occur between epochs need to be done on an internal copy of the tree while lookups need to be done on ``curSTR.tree``. Otherwise, the server has to generate an STR each time a new leaf is inserted into ``curSTR.tree`` for lookups to work as intended. I think it's more work to keep track of a shadow copy of the tree between epochs than to clone the tree first and then insert all changes.